tokio-util/src/sync/task_tracker.rs

```diff
@@ -0,0 +1,591 @@
+///     tracker.wait().await;
+///
+///     println!("This is printed after all of the tasks.");
+/// }
+/// ```
+///
+/// ## Wait for tasks to exit
+///
+/// This example shows the intended use-case of `TaskTracker`. It is used together with
+/// [`CancellationToken`] to implement graceful shutdown.
+/// ```
+/// use tokio_util::sync::{CancellationToken, TaskTracker};
+/// use tokio::time::{self, Duration};
+///
+/// async fn background_task(num: u64) {
+///     for i in 0..10 {
+///         time::sleep(Duration::from_millis(100*num)).await;
+///         println!("Background task {} in iteration {}.", num, i);
+///     }
+/// }
+///
+/// #[tokio::main]
+/// # async fn _hidden() {}
+/// # #[tokio::main(flavor = "current_thread", start_paused = true)]
+/// async fn main() {
+///     let tracker = TaskTracker::new();
+///     let token = CancellationToken::new();
+///
+///     for i in 0..10 {
+///         let token = token.clone();
+///         tracker.spawn(async move {
+///             // Use a `tokio::select!` to kill the background task if the token is
+///             // cancelled.
+///             tokio::select! {
+///                 () = background_task(i) => {
+///                     println!("Task {} exiting normally.", i);
+///                 },
+///                 () = token.cancelled() => {
+///                     // Do some cleanup before we really exit.
+///                     time::sleep(Duration::from_millis(50)).await;
+///                     println!("Task {} finished cleanup.", i);
+///                 },
+///             }
+///         });
+///     }
+///
+///     // Spawn a background task that will send the shutdown signal.
+///     {
+///         let tracker = tracker.clone();
+///         tokio::spawn(async move {
+///             // Normally you would use something like ctrl-c instead of
+///             // sleeping.
+///             time::sleep(Duration::from_secs(2)).await;
+///             tracker.close();
+///             token.cancel();
+///         });
+///     }
+///
+///     // Wait for all tasks to exit.
+///     tracker.wait().await;
+///
+///     println!("All tasks have exited now.");
+/// }
+/// ```
+///
+/// [`CancellationToken`]: crate::sync::CancellationToken
+/// [`JoinHandle::is_finished`]: tokio::task::JoinHandle::is_finished
+/// [`JoinSet`]: tokio::task::JoinSet
+/// [`close`]: Self::close
+/// [`join_next`]: tokio::task::JoinSet::join_next
+/// [`wait`]: Self::wait
+/// [graceful shutdown]: https://tokio.rs/tokio/topics/shutdown
+#[derive(Clone)]
+pub struct TaskTracker {
+    inner: Arc<TaskTrackerInner>,
+}
+
+/// Represents a task tracked by a [`TaskTracker`].
+#[must_use]
+#[derive(Debug)]
+pub struct TaskTrackerToken {
+    task_tracker: TaskTracker,
+}
+
+struct TaskTrackerInner {
+    /// Keeps track of the state.
+    ///
+    /// The lowest bit is whether the task tracker is closed.
+    ///
+    /// The rest of the bits count the number of tracked tasks.
```

Comment:
would it have been possible to use the `Arc`'s reference count for this, rather than essentially introducing a second reference count? i suppose this might be difficult since we would want the TaskTracker handle to also keep the shared state alive, so maybe not...

Reply:
I did consider that, but unfortunately it is not possible. The `TaskTracker` can be cloned, so you need two counters to separate them.

Another advantage of having it together with the closed bit is that it's possible to get consistent snapshots of both values together. Imagine dropping the last task and setting the closed bit concurrently. We need one of them to perform the wakeup, which is easy here because they both change the same atomic. With several atomics, it's unclear how to do this.

Reply:
yup, that makes sense. i think the current design is great, but was just curious if you'd explored that avenue as well. thanks!

Reply:
One thing you technically could do is have all tokens take up just one refcount of the `Arc`, and release that refcount when the token counter drops to zero. This would reduce the number of atomic operations to create a token. But this gets rather unsafe, and I didn't think it was a great idea.